src/Inference/LW.hs

```diff
@@ -1,54 +1,49 @@
+import           Control.Carrier.Draw         (DrawC, runDraw)
+import           Control.Carrier.LPTracer     (LPTracerC, runLPTracer)
+import           Control.Carrier.ObsReader    (ObsReaderC, runObsReader)
+import           Control.Carrier.SampleTracer (SampleTracerC, runSampleTracer)
+import           Control.Effect.Dist          (Dist (Dist))
+import           Control.Monad                (replicateM)
+import qualified Data.Map                     as Map (filterWithKey, foldr,
+                                                      (!?))
+import           Data.Maybe                   (isJust)
+import           Env                          (Env)
+import           Inference.SIM                (DistC, runDist)
+import           PrimDist                     (logProb)
+import           Sampler                      (Sampler)
+import           Trace                        (FromSTrace (..), STrace)
+import Control.Carrier.Lift (LiftC, runM)
 
 -- | Top-level wrapper for Likelihood-Weighting (LW) inference
-lw :: (FromSTrace env, es ~ '[ObsReader env, Dist, State STrace, Observe, Sample])
+lw :: (FromSTrace env)
     => Int            -- ^ number of LW iterations
-    -> Model env es a -- ^ model
     -> Env env        -- ^ model environment
+    -> ObsReaderC env (DrawC (SampleTracerC (LPTracerC (DistC (LiftC Sampler))))) a -- ^ model
     -> Sampler [(Env env, Double)] -- ^ [(output model environment, likelihood-weighting)]
-lw n model env = do
-  lwTrace <- replicateM n (runLW model env)
+lw n env model = do
+  lwTrace <- replicateM n (runLW env model)
   return $ map (\((_, strace), p) -> (fromSTrace strace, p)) lwTrace
 
 -- | Handler for one iteration of LW
-runLW :: es ~ '[ObsReader env, Dist, State STrace, Observe, Sample]
-  => Model env es a -- ^ model
-  -> Env env        -- ^ model environment
-  -> Sampler ((a, STrace), Double) -- ^ ((model output, sample trace), likelihood-weighting)
-runLW env = handleSamp . handleObs 0 . handleState Map.empty . traceSamples . handleCore env
-
--- | Handle each @Observe@ operation by computing and accumulating a log probability
-handleObs :: Member Sample es
-  => Double -- ^ accumulated log-probability
-  -> Prog (Observe : es) a
-  -> Prog es (a, Double) -- ^ (model output, final log-probability)
-handleObs logp (Val x) = return (x, exp logp)
-handleObs logp (Op u k) = case discharge u of
-    Right (Observe d y α) -> do
-      let logp' = logProb d y
-      handleObs (logp + logp') (k y)
-    Left op' -> Op op' (handleObs logp . k)
+runLW ::
+     Env env        -- ^ model environment
+  -> ObsReaderC env (DrawC (SampleTracerC (LPTracerC (DistC (LiftC Sampler))))) a -- ^ model
+  -> Sampler ((a, STrace), Double) -- ^ ((model output, zsample trace), likelihood-weighting)
+runLW env m = do
+    ((output, strace), lptrace) <- runM $ runDist $ runLPTracer False $ runSampleTracer $ runDraw $ runObsReader env m
+    let l = exp $ Map.foldr (+) 0 lptrace
```

Comment:
How come the old `runLW` didn't do this maths? Did it used to get done in handleObs?

Reply:
yup